Extension/lib/content-script/websocket.js

```diff
@@ -0,0 +1,296 @@
+                get: function () {
+                    return fallthruGet(this, 'bufferedAmount', 0);
+                },
+                set: noopfn
+            },
+            'extensions': {
+                get: function () {
+                    return fallthruGet(this, 'extensions', '');
+                },
+                set: noopfn
+            },
+            'onclose': {
+                get: function () {
+                    return fallthruGet(this, 'onclose', null);
+                },
+                set: function (value) {
+                    fallthruSet(this, 'onclose', value);
+                }
+            },
+            'onerror': {
+                get: function () {
+                    return fallthruGet(this, 'onerror', null);
+                },
+                set: function (value) {
+                    fallthruSet(this, 'onerror', value);
+                }
+            },
+            'onmessage': {
+                get: function () {
+                    return fallthruGet(this, 'onmessage', null);
+                },
+                set: function (value) {
+                    fallthruSet(this, 'onmessage', value);
+                }
+            },
+            'onopen': {
+                get: function () {
+                    return fallthruGet(this, 'onopen', null);
+                },
+                set: function (value) {
+                    fallthruSet(this, 'onopen', value);
+                }
+            },
+            'protocol': {
+                get: function () {
+                    return fallthruGet(this, 'protocol', '');
+                },
+                set: noopfn
+            },
+            'readyState': {
+                get: function () {
+                    return fallthruGet(this, 'readyState', 0);
+                },
+                set: noopfn
+            },
+            'url': {
+                get: function () {
+                    return fallthruGet(this, 'url', '');
+                },
+                set: noopfn
+            }
+        });
+        toWrapped.set(this, {
+            args: {url: url, protocols: protocols},
+            listeners: [],
+            properties: {}
+        });
+
+        var that = this;
+
+        function messageListener(event) {
+            if (!(event.data.direction &&
+                event.data.direction == "to-page-script" &&
+                event.data.elementUrl == url &&
+                event.data.documentUrl == document.URL)) {
+                return;
+            }
+
+            onResponseReceived(that, !event.data.collapse)
+        }
+
+        window.addEventListener("message", messageListener, false);
+
+        var message = {
+            direction: 'from-page-script',
+            elementUrl: url,
+            documentUrl: document.URL
+        };
+
+        // Send a message to the background page to check if the request should be blocked
+        window.postMessage(message, "*");
```

Comment:
Both postMessage and addEventListener may be tampered from a page code.
Save both in a local variables at the very start of the script and call it like that:
postMessage.call(window, parameters)